tensorflow_addons/layers/multihead_attention.py

```diff
@@ -0,0 +1,253 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+import typing
+
+import tensorflow as tf
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class MultiHeadAttention(tf.keras.layers.Layer):
+    r"""
+    MultiHead Attention layer.
+
+    Defines the MultiHead Attention operation as defined in [Attention Is All You Need](https://arxiv.org/abs/1706.03762) which takes in a `query`, `key` and `value` tensors returns the dot-product attention between them:
+
+        ```python
+        mha = MultiHeadAttention(head_size=128, num_heads=128)
+
+        query = tf.random.uniform((32, 20, 200)) # (B, N, Dq)
+        key = tf.random.uniform((32, 15, 300)) # (B, M, Dk)
+        value = tf.random.uniform((32, 15, 400)) # (B, M, Dv)
+
+        attention = mha([query, key, value]) # (B, N, Dv)
```

Comment:
I think keras layer supports keyword argument now, so you don't necessary to pass in all the values as a list.

Reply:
I wanted to do that but `input_shapes` in `build` is still a list of shape of the first argument in `call`, if there is another way of passing the shapes of the other arguments to `build` I'd like to implement it like that.

Reply:
Ah, good point. Then we I think the current approach is probably the only way.